### PR TITLE
Added `Tensor.to` overloads to `torch._refs.to`

### LINKED ISSUE #84264: Support more overloads of Tensor.to in torch._refs.to
### 🚀 The feature, motivation and pitch

There are multiple overloads of `to`, but currently only one overload is supported in `torch._refs.to`
aten/src/ATen/native/native_functions.yaml lists all overloads:
https://github.com/pytorch/pytorch/blob/acd6ca8cfa9537284928fb5d36834d1e5ae1e6f3/aten/src/ATen/native/native_functions.yaml#L6581-L6599
Missing overloads:

- [ ]  to.dtype_layout
- [ ]  to.device
- [ ]  to.other

### Alternatives

_No response_

### Additional context

_No response_

cc @ezyang @mruberry @ngimel @jjsjann123

### PR DESCRIPTION
**Pull request by jjsjann123**
Fixes #84264

### CODE REVIEW

**Comment by jjsjann123** [comment]
This should replace #84735

File: test/test_prims.py
Comment (IvanYashchuk) [inline test/test_prims.py:313]: Let's use OpInfo for tests instead, then this would be tested in `test_ops.py`. We need to uncomment this:
https://github.com/pytorch/pytorch/blob/c5a8946e40d6cda42aa38dda2705ea4e9930c2cb/torch/testing/_internal/common_methods_invocations.py#L15984-L15989

Comment (jjsjann123) [reply test/test_prims.py]: Thx for suggesting the opinfo test. Turns out my python overload is buggy :smile:

Comment (jjsjann123) [reply test/test_prims.py]: I'm actually also able to repro the data dependent control flow issue~~ double win~

Comment (jjsjann123) [reply test/test_prims.py]: oh... the control flow issue went away after this
```
-            or a.is_contiguous(memory_format=memory_format)
+            or utils.is_contiguous_for_memory_format(a, memory_format=memory_format)
```

File: torch/_refs/__init__.py
Comment (IvanYashchuk) [inline torch/_refs/__init__.py:1723]: It's unfortunate that `to.device` overload takes 6 positional arguments and `to.dtype`, `to.other` take 5 positional arguments.

File: torch/_refs/__init__.py
Comment (IvanYashchuk) [inline torch/_refs/__init__.py:1686]: layout and pin_memory are not modified in this function, maybe we shouldn't pass it?

File: torch/_refs/__init__.py
Comment (IvanYashchuk) [reply torch/_refs/__init__.py:1723]: The `sample_inputs_to` function I added is not testing that.
```
diff --git a/torch/testing/_internal/common_methods_invocations.py b/torch/testing/_internal/common_methods_invocations.py
index 9a9d8c7a62..8b2e0b2834 100644
--- a/torch/testing/_internal/common_methods_invocations.py
+++ b/torch/testing/_internal/common_methods_invocations.py
@@ -3898,31 +3898,16 @@ def sample_inputs_to(op_info, device, dtype, requires_grad, **kwargs):

     # to.device overload
     for device, nb, cp, mem_f in product(devices, [True, False], [True, False], memory_formats):
-        kwargs = {
-            "non_blocking": nb,
-            "copy": cp,
-            "memory_format": mem_f,
-        }
-        yield SampleInput(make_arg((S, S, S, S)), args=(device, torch.float64,), kwargs=kwargs)
+        yield SampleInput(make_arg((S, S, S, S)), args=(device, torch.float64, nb, cp, mem_f))

     # to.dtype overload
     for nb, cp, mem_f in product([True, False], [True, False], memory_formats):
-        kwargs = {
-            "non_blocking": nb,
-            "copy": cp,
-            "memory_format": mem_f,
-        }
-        yield SampleInput(make_arg((S, S, S, S)), args=(torch.float64,), kwargs=kwargs)
+        yield SampleInput(make_arg((S, S, S, S)), args=(torch.float64, nb, cp, mem_f))

     # to.other overload
     for device, nb, cp, mem_f in product(devices, [True, False], [True, False], memory_formats):
-        kwargs = {
-            "non_blocking": nb,
-            "copy": cp,
-            "memory_format": mem_f,
-        }
         other = make_arg((S, S, S, S), dtype=torch.float64, device=device)
-        yield SampleInput(make_arg((S, S, S, S)), args=(other,), kwargs=kwargs)
+        yield SampleInput(make_arg((S, S, S, S)), args=(other, nb, cp, mem_f))
```

File: torch/_refs/__init__.py
Comment (IvanYashchuk) [inline torch/_refs/__init__.py:1750]:
```suggestion
        memory_format or torch.preserve_format
```

**Review by IvanYashchuk — COMMENTED** [review]
The case for which there were no tests is that `to.device` accepts 6 positional arguments, while the other two overloads accept 5. For robustness, we need to support that extra arg.

File: torch/_refs/__init__.py
Comment (mruberry) [inline torch/_refs/__init__.py:1658]: This function needs a comment describing its purpose and what it does

File: torch/_refs/__init__.py
Comment (mruberry) [inline torch/_refs/__init__.py:1741]: Let's add a comment here that this is a short-circuit for when just a dtype change is requested.

Seems like this short-circuit could be more aggressive?

File: torch/_refs/__init__.py
Comment (mruberry) [reply torch/_refs/__init__.py:1686]: I see `layout` is used below, but it does seem like `pin_memory` isn't (because we don't have any prims that support it?)

Comment (mruberry) [reply torch/_refs/__init__.py:1686]: Instead of removing unused params from the signature, let's throw an unsupported error when they're passed

File: torch/_refs/__init__.py
Comment (mruberry) [reply torch/_refs/__init__.py:1750]: I prefer explicitly checking for `None` most of the time because lots of Python objects evaluate to `False`

File: torch/_refs/__init__.py
Comment (mruberry) [inline torch/_refs/__init__.py:1725]: We should add a comment about our `.to` parsing strategy. `to` is a tricky and kinda weird function. What about making the ref have the signature:

```
def to(a: TensorLikeType, *args, **kwargs)
```

and then inspecting the first element of `*args` (if it exists) to see if it's a device, dtype, or tensor? Then we can extract the properties, diff them against the tensor's current metadata to come back with a subset of properties that need to change, and from there call `empty_like` or short-circuit to just a `dtype` or `device` conversion as appropriate?

**Review by mruberry — COMMENTED** [review]
Overall looks pretty good, @jjsjann123! Maybe a few opportunities to refine the `to` parsing, which is super tricky, and add more comments

File: torch/_refs/__init__.py
Comment (jjsjann123) [reply torch/_refs/__init__.py:1725]: ^^^ good point. In terms of implementation, I think it's gonna look roughly about the same here. But the intention of dispatch would be cleaner comparing to swapping arguments in current `_to_dispatch`

File: torch/_refs/__init__.py
Comment (jjsjann123) [reply torch/_refs/__init__.py:1741]: Will update the short-circuit condition~

File: torch/_refs/__init__.py
Comment (jjsjann123) [reply torch/_refs/__init__.py:1723]: To be fair, I don't think we have comprehensive tests after the changes neither. There could also be combinations of positional + kw args, which affects how dispatching in python is handling this. hahaha.

But I agree that, knowing that the dispatch is more bug prone on mapping positional arguments, changes here to the tests looks like a good idea. Will update.

File: torch/_refs/__init__.py
Comment (jjsjann123) [reply torch/_refs/__init__.py:1725]: Lol, after I started writing this I realized that we'll have `args` and `kwargs` possibly overlapping. so we'll need more checks for that. :vomiting_face:

**Comment by mruberry** [comment]
@ngimel suggests we should add a layout swapping prim (not in this PR) and that it may help us avoid the `empty_like` and `copy_to` construction, which is an inplace modification we're hoping to avoid

When a layout changing prim is added then this code can be updated, so no change is needed for this PR, I'm just noting the idea here

**Comment by jjsjann123** [comment]
> @ngimel suggests we should add a layout swapping prim (not in this PR) and that it may help us avoid the `empty_like` and `copy_to` construction, which is an inplace modification we're hoping to avoid
> 
> When a layout changing prim is added then this code can be updated, so no change is needed for this PR, I'm just noting the idea here

Sounds good. We also are lacking support on `non_blocking` in prim. Which I noted as a todo here in this PR. We can clean it up in follow up as well.

**Comment by jjsjann123** [comment]
We seem to needed some extra bits support from fake tensor on `is_pinned`

```
======================================================================
FAIL: test_python_ref_meta__refs_to_cuda_uint8 (__main__.TestCommonCUDA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/pytorch/pytorch/torch/testing/_internal/common_utils.py", line 1950, in wrapper
    method(*args, **kwargs)
  File "/opt/pytorch/pytorch/torch/testing/_internal/common_utils.py", line 1950, in wrapper
    method(*args, **kwargs)
  File "/opt/pytorch/pytorch/torch/testing/_internal/common_device_type.py", line 378, in instantiated_test
    result = test(self, **param_kwargs)
  File "/opt/pytorch/pytorch/torch/testing/_internal/common_device_type.py", line 815, in test_wrapper
    return test(*args, **kwargs)
  File "/opt/pytorch/pytorch/torch/testing/_internal/common_device_type.py", line 1011, in only_fn
    return fn(self, *args, **kwargs)
  File "test/test_ops.py", line 187, in test_python_ref_meta
    meta_result = op(meta_sample.input, *meta_sample.args, **meta_sample.kwargs)
  File "/opt/pytorch/pytorch/torch/testing/_internal/opinfo/core.py", line 1026, in __call__
    return self.op(*args, **kwargs)
  File "/opt/pytorch/pytorch/torch/_refs/__init__.py", line 1751, in to
    device, dtype, non_blocking, copy, memory_format, layout, pin_memory = _to_dispatch(
  File "/opt/pytorch/pytorch/torch/_refs/__init__.py", line 1700, in _to_dispatch
    pin_memory = args0.is_pinned()
  File "/opt/pytorch/pytorch/torch/utils/_python_dispatch.py", line 74, in wrapped
    return f(self, *args, **kwargs)
  File "/opt/pytorch/pytorch/torch/_subclasses/fake_tensor.py", line 850, in __torch_dispatch__
    assert func not in _device_not_kwarg_ops, f"NYI: {func}"
AssertionError: NYI: aten.is_pinned.default
```

File: torch/_refs/__init__.py
Comment (jjsjann123) [reply torch/_refs/__init__.py]: O.o Looks like the `to.device` is not exposed in python. Which seems strange as native_functions.yaml seems to have that entry. But I also don't know how aten handles overload/dispatch in python, I guess if frontend doesn't have that, it's safe to just not test it :rofl: 

```
In [3]: x.to("cuda:0", torch.float16, False, False, None)
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Input In [3], in <cell line: 1>()
----> 1 x.to("cuda:0", torch.float16, False, False, None)

TypeError: to() takes from 0 to 4 positional arguments but 5 were given
```

**Comment by jjsjann123** [comment]
I think I addressed all comments. Fingers crossed that lint will also be good this time. :crossed_fingers:

**Comment by IvanYashchuk** [comment]
We still need to update the sample_inputs function because the current version doesn't test all positional arguments https://github.com/pytorch/pytorch/pull/84802#discussion_r968258574.

**Comment by jjsjann123** [comment]
> We still need to update the sample_inputs function because the current version doesn't test all positional arguments [#84802 (comment)](https://github.com/pytorch/pytorch/pull/84802#discussion_r968258574).

Just updated it to use positional for most things except `memory_format`, which only takes kwarg in python API: https://github.com/pytorch/pytorch/issues/84265#issuecomment-1245647430

**Review by IvanYashchuk — APPROVED** [review]

**Comment by IvanYashchuk** [comment]
@mruberry could you please take a final look?

File: torch/_refs/__init__.py
Comment (mruberry) [inline torch/_refs/__init__.py:1661]: Nice comment

File: torch/_refs/__init__.py
Comment (mruberry) [inline torch/_refs/__init__.py:1722]: Add a comment for each overload with what the signature can be

File: torch/_refs/__init__.py
Comment (mruberry) [inline torch/_refs/__init__.py:1725]: Update the asserts to use the `check` function from utils

File: torch/_refs/__init__.py
Comment (mruberry) [inline torch/_refs/__init__.py:1713]: This pattern seems fine. One easier alternative might be to define helper functions like `_to_tensor` with the correct signature and pass `*args` and `**kwargs` to them and have them return a dict of the correct metadata

Comment (mruberry) [reply torch/_refs/__init__.py:1713]: (If anyone ever has to refactor this code, that's the approach I imagine they'll take)

Comment (mruberry) [reply torch/_refs/__init__.py:1713]: (It might be interesting to look at how `bind` is used here:

https://github.com/pytorch/pytorch/blob/706b99030656c573619cebaa3be9298a575fc776/torch/_prims_common/wrappers.py#L99

)

File: torch/_refs/__init__.py
Comment (mruberry) [inline torch/_refs/__init__.py:1764]: Nice comment

File: torch/_refs/__init__.py
Comment (mruberry) [inline torch/_refs/__init__.py:1766]: This approach is fine for now but it'll be tricky to maintain. It'd be nicer if the changing properties were listed separately. Maybe `_to_dispatch` should return two dicts, and one of them is what's mutating? Maybe it should return only what's changing? Does the call to `empty_like` have to specify what isn't changing, to, or would it inherit that metadata, anyway?

File: torch/_refs/__init__.py
Comment (mruberry) [inline torch/_refs/__init__.py:1776]: Seems like this should go earlier, before `memory_format` is queried as part of the test for whether this can short-circuit to `prims.convert_element_type`? Maybe it makes more sense to have `memory_format` set properly from `_to_dispatch`?

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:3898]: Nice comment

File: torch/testing/_internal/common_methods_invocations.py
Comment (mruberry) [inline torch/testing/_internal/common_methods_invocations.py:16005]: Cool!

**Review by mruberry — APPROVED** [review]
Cool! I made some inline comments for your review, @jjsjann123. I encourage you to consider some refactoring so the code is more maintainable, but it's not required, and could also be separated to a future PR for code velocity. 

There are some inline comments that should be addressed before this is merged, however, like using `check` instead of `assert`, and ensuring the setting of `memory_format` happens in the correct place.

File: torch/_refs/__init__.py
Comment (jjsjann123) [reply torch/_refs/__init__.py:1713]: ! Sorry about my sloppy python~
I was sure how to handle kwargs binding cleanly in my earlier attempt so ended up with them explicit in the signature. I'll make another attempt to switch to inspect.signature.

Comment (jjsjann123) [reply torch/_refs/__init__.py:1713]: I need to think a little bit more about this. If we can magically have all arguments converted to a proper dictionary, then we would not have to worry about doing this manual~ish dispatch.

But the tricky part is that each of `torch.ops.aten.to.XXX` overload takes arguments slightly differently and they are implicit in the sense that they all takes `*args, **kwargs` and `inspect` can't really do anything useful there.

> One easier alternative might be to define helper functions like _to_tensor with the correct signature

^^^ I don't know how we can have one `_to_tensor` signature and resolve the `*args`. To figure out the potential `*args` dispatch, I think we'll need three `_to_tensor_XXX` variants and use that with `inspect` to automatically sort out the signature. But that seems to bring us the full circle back to the original implementation, where we are just if/elif on the input arguments. :laughing: 

Maybe I'm missing something obvious, I can use some more elaborated comments. :bow:

Comment (IvanYashchuk) [reply torch/_refs/__init__.py:1713]: Here's one suggestion on how to handle dispatch. At least one positional argument must be passed to `_to_dispatch` in this example but everything else can be passed as either a positional or a keyword argument.

```py
import torch
from functools import singledispatch

@singledispatch
def _to_dispatch(*args, **kwargs):
    raise NotImplementedError

@_to_dispatch.register
def _to_device(device: torch.device, dtype, non_blocking, copy, memory_format):
    print("to.device overload!")
    layout = None
    pin_memory = None
    return device, dtype, non_blocking, copy, memory_format, layout, pin_memory

@_to_dispatch.register
def _to_device(device: str, dtype, non_blocking, copy, memory_format):
    print("to.device str overload!")
    device = torch.device(device)
    return _to_dispatch(device, dtype, non_blocking, copy, memory_format)

@_to_dispatch.register
def _to_dtype(dtype: torch.dtype, non_blocking, copy, memory_format):
    print("to.dtype overload!")
    layout = None
    pin_memory = None
    device = None
    return device, dtype, non_blocking, copy, memory_format, layout, pin_memory

@_to_dispatch.register
def _to_other(other: torch.Tensor, non_blocking, copy, memory_format):
    print("to.other overload!")
    layout = other.layout
    pin_memory = other.is_pinned()
    device = other.device
    dtype = other.dtype
    return device, dtype, non_blocking, copy, memory_format, layout, pin_memory
```

Comment (jjsjann123) [reply torch/_refs/__init__.py]: Thanks a lot for hand-holding me through this :smile: 

I went with singledispatch that returns a dictionary for kwargs. I feel this is a relatively straightforward way to handle this messy overload. Feels like this also works more naturally with the suggestion @mruberry has on canonicalize the kwargs per attributes from input tensor.

File: torch/_refs/__init__.py
Comment (jjsjann123) [reply torch/_refs/__init__.py]: I think `empty_like` inherits whatever was from the original tensor. I added `canonicalize_to_arguments` that examines kwargs and attributes from input tensor and update `kwargs` to contain only what's changing.

Hopefully this makes it cleaner on the short-cut condition here.

**Comment by jjsjann123** [comment]
Many thanks for the suggestion on handling overload in python. I've updated that section and went with a combination of `singledispatch` on positional arguments + returned dictionary.

Addressed other review comments as well. This PR should be good for another round of review.

**Comment by jjsjann123** [comment]
Lol. clicking `request for review` seems to have invalidated the approval stamp from earlier?!

**Pull request closed without merging**
